internal/controller/controllers/kube_compat.go

```diff
@@ -2,19 +2,195 @@ package controllers
+	is := &certtypes.Issuer{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      selfSignedIssuerName,
+			Namespace: asc.namespace,
+		},
+	}
+	mutateFn := func() error {
+		if err := controllerutil.SetControllerReference(asc.Object, is, asc.rec.Scheme); err != nil {
+			return err
+		}
+		is.Spec = certtypes.IssuerSpec{
+			IssuerConfig: certtypes.IssuerConfig{
+				SelfSigned: &certtypes.SelfSignedIssuer{},
+			},
+		}
+		return nil
+	}
+	return is, mutateFn, nil
+}
+
+// newCACert describes how to create a Certificate for the root CA
+func newCACert(ctx context.Context, log logrus.FieldLogger, asc ASC) (client.Object, controllerutil.MutateFn, error) {
+	cert := &certtypes.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      caIssuerName,
+			Namespace: asc.namespace,
+		},
+	}
+
+	mutateFn := func() error {
+		if err := controllerutil.SetControllerReference(asc.Object, cert, asc.rec.Scheme); err != nil {
+			return err
+		}
+		cert.Spec.CommonName = caCommonName
+		cert.Spec.IsCA = true
+		cert.Spec.SecretName = caIssuerName
+		cert.Spec.IssuerRef = certmeta.ObjectReference{
+			Kind: "Issuer",
+			Name: selfSignedIssuerName,
+		}
+		return nil
+	}
+
+	return cert, mutateFn, nil
+}
+
+// newCAIssuer describes how to create an Issuer for creating certificates signed by the root CA cert
+func newCAIssuer(ctx context.Context, log logrus.FieldLogger, asc ASC) (client.Object, controllerutil.MutateFn, error) {
+	is := &certtypes.Issuer{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      caIssuerName,
+			Namespace: asc.namespace,
+		},
+	}
+	mutateFn := func() error {
+		if err := controllerutil.SetControllerReference(asc.Object, is, asc.rec.Scheme); err != nil {
+			return err
+		}
+		is.Spec = certtypes.IssuerSpec{
+			IssuerConfig: certtypes.IssuerConfig{
+				CA: &certtypes.CAIssuer{
+					SecretName: caIssuerName,
+				},
+			},
+		}
+		return nil
+	}
+
+	return is, mutateFn, nil
+}
+
+// newCertificate describes how to create a Certificate for a service signed by the assisted installer CA
+func newCertificate(serviceName string, asc ASC) (client.Object, controllerutil.MutateFn, error) {
+	cert := &certtypes.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceName,
+			Namespace: asc.namespace,
+		},
+	}
+
+	mutateFn := func() error {
+		if err := controllerutil.SetControllerReference(asc.Object, cert, asc.rec.Scheme); err != nil {
+			return err
+		}
+		cert.Spec.SecretName = serviceName
+		cert.Spec.IssuerRef = certmeta.ObjectReference{
+			Kind: "Issuer",
+			Name: caIssuerName,
+		}
+		cert.Spec.DNSNames = []string{
+			fmt.Sprintf("%s.%s.svc", serviceName, asc.namespace),
+			fmt.Sprintf("%s.%s.svc.cluster.local", serviceName, asc.namespace),
+		}
+		return nil
+	}
+
+	return cert, mutateFn, nil
+}
+
+func newWebhookCert(ctx context.Context, log logrus.FieldLogger, asc ASC) (client.Object, controllerutil.MutateFn, error) {
```

Comment:
Same here, does `ctx context.Context, log logrus.FieldLogger` need to be in the argument list? :o

Reply:
Yes, because this is a [`NewComponentFn`](https://github.com/openshift/assisted-service/blob/d2bba26c003c543612de88883f936d54aceb4ca2/internal/controller/controllers/agentserviceconfig_controller.go#L187)